From d6ea5b25bbae140b9fecfd4eefba23ed25b1728a Mon Sep 17 00:00:00 2001
From: Matthias Brugger <mbrugger@suse.com>
Date: Wed, 19 Apr 2017 19:02:37 +0200
Subject: [PATCH] net: hns: set relation between hns_dsaf and hns_mdio
References: bsc#1033895
Patch-mainline: never, this is a temporary fix of the probe order

hns_dsaf relies on loading hns_mdio first. We hack a fixed call relation
which will fix module load and unload.

Signed-off-by: Matthias Brugger <mbrugger@suse.com>
---
 drivers/net/ethernet/hisilicon/hns/hns_dsaf_main.c |  5 +++++
 drivers/net/ethernet/hisilicon/hns_mdio.c          | 12 ++++++++++++
 drivers/net/ethernet/hisilicon/hns_mdio.h          |  7 +++++++
 3 files changed, 24 insertions(+)
 create mode 100644 drivers/net/ethernet/hisilicon/hns_mdio.h

diff --git a/drivers/net/ethernet/hisilicon/hns/hns_dsaf_main.c b/drivers/net/ethernet/hisilicon/hns/hns_dsaf_main.c
index e0bc79ea3d88..f39a49edb5fd 100644
--- a/drivers/net/ethernet/hisilicon/hns/hns_dsaf_main.c
+++ b/drivers/net/ethernet/hisilicon/hns/hns_dsaf_main.c
@@ -27,6 +27,7 @@
 #include "hns_dsaf_ppe.h"
 #include "hns_dsaf_rcb.h"
 #include "hns_dsaf_misc.h"
+#include "../hns_mdio.h"
 
 const char *g_dsaf_mode_match[DSAF_MODE_MAX] = {
 	[DSAF_MODE_DISABLE_2PORT_64VM] = "2port-64vf",
@@ -2738,6 +2739,8 @@ static int hns_dsaf_probe(struct platform_device *pdev)
 		return ret;
 	}
 
+	hns_mdio_register();
+
 	ret = hns_dsaf_get_cfg(dsaf_dev);
 	if (ret)
 		goto free_dev;
@@ -2793,6 +2796,8 @@ static int hns_dsaf_remove(struct platform_device *pdev)
 
 	hns_dsaf_free_dev(dsaf_dev);
 
+	hns_mdio_unregister();
+
 	return 0;
 }
 
diff --git a/drivers/net/ethernet/hisilicon/hns_mdio.c b/drivers/net/ethernet/hisilicon/hns_mdio.c
index 501eb2090ca6..7de3efa5d17b 100644
--- a/drivers/net/ethernet/hisilicon/hns_mdio.c
+++ b/drivers/net/ethernet/hisilicon/hns_mdio.c
@@ -541,6 +541,18 @@ static int hns_mdio_probe(struct platform_device *pdev)
 	return 0;
 }
 
+void hns_mdio_register(void)
+{
+	__module_get(THIS_MODULE);
+}
+EXPORT_SYMBOL(hns_mdio_register);
+
+void hns_mdio_unregister(void)
+{
+	module_put(THIS_MODULE);
+}
+EXPORT_SYMBOL(hns_mdio_unregister);
+
 /**
  * hns_mdio_remove - remove mdio device
  * @pdev: mdio platform device
diff --git a/drivers/net/ethernet/hisilicon/hns_mdio.h b/drivers/net/ethernet/hisilicon/hns_mdio.h
new file mode 100644
index 000000000000..4d98211961a2
--- /dev/null
+++ b/drivers/net/ethernet/hisilicon/hns_mdio.h
@@ -0,0 +1,7 @@
+#ifndef __HNS_MDIO_H
+#define __HNS_MDIO_H
+
+void hns_mdio_register(void);
+void hns_mdio_unregister(void);
+
+#endif
-- 
2.12.0

